### PR TITLE
[Feat] support qwen2_5 _vl in multimodal framework

### PR DESCRIPTION
**Pull request by pengchengneo**
as title ,still in dev

this PR is adapted from 

https://github.com/sgl-project/sglang-jax/pull/652
https://github.com/sgl-project/sglang-jax/pull/633

## Test Verification Summary

launch command

```
uv run python3 -u -m sgl_jax.launch_server --multimodal --model-path Qwen/Qwen2.5-VL-3B-Instruct --log-requests --disable-precompile
```

### 1. Test Assets (Input)

| Image 1 | Image 2 | Video |
| :---: | :---: | :---: |
| Image 1 | Image 2 | <video src="https://github.com/user-attachments/assets/4efbe8b1-b1e2-4b01-9472-f7cd3e6a7cb0" width="200"></video> |

### 2. Inference Support Matrix

The following table summarizes the verification of Qwen2.5-VL models across different sizes and hardware configurations.

| Model Name | Test Command (Request) | Inference Result (Response) | Status |
| :--- | :--- | :--- | :---: |
| **Qwen2.5-VL-3B-Instruct** | <details><summary><b>View curl</b></summary><pre>curl -X POST "http://localhost:30000/v1/chat/completions" \<br> -H "Content-Type: application/json" \<br> -d '{<br> "model": "/models/Qwen/Qwen2.5-VL-3B-Instruct",<br> "messages": [<br> {<br> "role": "user",<br> "content": [<br> {"type": "text", "text": "介绍下图片的差异"},<br> {"type": "image_url", "image_url": {"url": "1.png"}},<br> {"type": "image_url", "image_url": {"url": "2.png"}}<br> ]<br> }<br> ],<br> "max_tokens": 300<br> }'</pre></details> | <details><summary><b>View JSON</b></summary><pre>{"id":"ec130079d83142b982bef9eb092ed2cd","object":"chat.completion","created":1769511300,"model":"/models/Qwen/Qwen2.5-VL-3B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"这张图片展示了一位冲浪者在海浪上骑行的场景。冲浪者穿着潜水衣，熟练地驾驭着冲浪板，身体的姿态显示出他正在骑行中。背景中可以看到蔚蓝色的天空和远处的山脉，山脉上覆盖着白色的积雪，显得非常壮丽。天空中还有一架飞机在飞行，为整个画面增添了一丝动感。冲浪者的动作和周围的自然景观共同构成了一个充满活力和动感的场景。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151645,"hidden_states":null}],"usage":{"prompt_tokens":1938,"total_tokens":2038,"completion_tokens":100,"prompt_tokens_details":null}}</pre></details> | ✅ |
| **Qwen2.5-VL-7B-Instruct** | <details><summary><b>View curl</b></summary><pre>curl -X POST "http://localhost:30000/v1/chat/completions" \<br> -H "Content-Type: application/json" \<br> -d '{<br> "model": "/models/Qwen/Qwen2.5-VL-7B-Instruct",<br> "messages": [<br> {<br> "role": "user",<br> "content": [<br> {"type": "text", "text": "介绍下图片的差异"},<br> {"type": "image_url", "image_url": {"url": "1.png"}},<br> {"type": "image_url", "image_url": {"url": "2.png"}}<br> ]<br> }<br> ],<br> "max_tokens": 300<br> }'</pre></details> | <details><summary><b>View JSON</b></summary><pre>{"id":"7e28c65cb98142d1832f313564fd3800","object":"chat.completion","created":1769510884,"model":"/models/Qwen/Qwen2.5-VL-7B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"The images provided are nearly identical, showing a surfer riding a wave. Here are the key differences:\n\n1. **Background Details:**\n - In the first image, there are mountains in the background, which are not present in the second image.\n - The second image focuses solely on the surfer and the wave, with a more limited view of the surrounding environment.\n\n2. **Color and Lighting:**\n - The first image has a broader range of colors, including the blue of the sky and the white of the clouds, which are not present in the second image.\n - The second image appears to have a more focused and warm-toned lighting, possibly due to a different time of day or a filter applied to the photo.\n\n3. **Airplane:**\n - In the first image, there is a small airplane flying in the sky to the left of the surfer.\n - The second image does not include the airplane.\n\n4. **Wave Details:**\n - The wave in the second image appears to be a bit more turbulent and dynamic, with more splashing water, compared to the first image.\n\nOverall, the main differences are the inclusion of mountains in the first image and the broader sky and cloud details, as well as the absence of the airplane in the second image.","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151645,"hidden_states":null}],"usage":{"prompt_tokens":1938,"total_tokens":2204,"completion_tokens":266,"prompt_tokens_details":null}}</pre></details> | ✅ |
| **Qwen2.5-VL-32B-Instruct** | <details><summary><b>View curl</b></summary><pre>curl -X POST "http://localhost:30000/v1/chat/completions" \<br> -H "Content-Type: application/json" \<br> -d '{<br> "model": "/models/Qwen/Qwen2.5-VL-32B-Instruct",<br> "messages": [<br> {<br> "role": "user",<br> "content": [<br> {"type": "text", "text": "介绍下图片的差异"},<br> {"type": "image_url", "image_url": {"url": "1.png"}},<br> {"type": "image_url", "image_url": {"url": "2.png"}}<br> ]<br> }<br> ],<br> "max_tokens": 300<br> }'</pre></details> | <details><summary><b>View JSON</b></summary><pre>{"id":"daff7f206e2e4c3485342d5f7f509ef6","object":"chat.completion","created":1769516602,"model":"/models/Qwen/Qwen2.5-VL-32B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"这两张图片虽然主题相似，但存在一些明显的差异。以下是两者的对比分析：\n\n---\n\n### **第一张图片的特点：**\n1. **背景：**\n - 背景中可以看到远处的雪山，天空晴朗，蓝色的天空中有一些云\n 。\n - 天空的颜色非常鲜明，给人一种开阔、清新的感觉。\n - 雪山的存在为画面增添了壮丽的自然景观。\n\n2. **光线：**\n - 光线明亮，色彩鲜艳，天空和海水的蓝色对比强烈。\n - 太阳似乎位于较高的位置，光线较为均匀。\n\n3. **人物姿态：**\n - 冲浪者的姿态稳定，身体稍微向后倾斜，似乎在控制平衡。\n - 冲浪板的姿态相对平缓，没有明显的颠簸。\n\n4. **飞机：**\n - 背景中有一架飞行\n 的飞机，飞行高度较高，与雪山形成对比。\n\n5. **海水：**\n - 海水的颜色偏深蓝色，波浪相对平缓，没有明显的白色泡沫。\n\n---\n\n### **第二张图片的特点：**\n1. **背景：**\n - 背景中没有雪山，而是更加接近海岸的场景。\n - 天空的颜色偏暖，可能是日出或日落时分，光线柔和。\n\n2. **光线：**\n - 光线柔和，带有一些金色或暖色调，给人一种","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"length","matched_stop":null,"hidden_states":null}],"usage":{"prompt_tokens":1938,"total_tokens":2238,"completion_tokens":300,"prompt_tokens_details":null}}</pre></details> | ✅ |
| **Qwen2.5-VL-72B-Instruct** | <details><summary><b>View curl</b></summary><pre>curl -X POST "http://localhost:30000/v1/chat/completions" \<br> -H "Content-Type: application/json" \<br> -d '{<br> "model": "/models/Qwen/Qwen2.5-VL-72B-Instruct",<br> "messages": [<br> {<br> "role": "user",<br> "content": [<br> {"type": "text", "text": "介绍下图片的差异"},<br> {"type": "image_url", "image_url": {"url": "1.png"}},<br> {"type": "image_url", "image_url": {"url": "2.png"}}<br> ]<br> }<br> ],<br> "max_tokens": 300<br> }'</pre></details> | <details><summary><b>View JSON</b></summary><pre>{"id":"122a55973f424eb9a7a4aac6ed34e719","object":"chat.completion","created":1769575926,"model":"/models/Qwen/Qwen2.5-VL-72B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"这两张图片展示了冲浪者的不同场景，以下是它们的主要差异：\n\n1. **背景**：\n - 第一张图片的背景是一个雪山和蓝天，远处还有一架飞行的飞机，给人一种壮丽的自然景观和冒险的感觉。\n - 第\n 张图片的背景是一个相对平坦的海面，没有明显的地标或飞行器，更加突出冲浪者的动作和海浪的细节。\n\n2. **光线和色彩**：\n - 第一张图片的光线较为柔和，色彩对比鲜明，雪山的白色和蓝天的蓝色形成了强烈的对比。\n - 第二张图片的光线更加明亮，色彩偏向于温暖的色调，海浪的绿色和冲浪者的黑色潜水服形成了鲜明的对比。\n\n3. **冲浪者的动作**：\n - 第一张图片中，冲浪者的动作更加夸张，似乎在\n 一个高难度的跳跃或转身动作，身体几乎垂直于水面。\n - 第二张图片中，冲浪者的动作相对稳定，似乎在做一个常规的转弯或骑行动作，身体与水面的角度较为平缓。\n\n4. **海浪的状态**：\n - 第一张图片中的海浪较大，浪花飞溅，给人一种动感和力量感。\n - 第二张图片中的海浪相对较小，浪花较少，更加突出冲浪者的技巧和控制力。\n\n总的来说，第一张图片更加突出自然景观和冒险的氛围，而第二张图片更\n 突出","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"length","matched_stop":null,"hidden_states":null}],"usage":{"prompt_tokens":1938,"total_tokens":2238,"completion_tokens":300,"prompt_tokens_details":null}}</pre></details> | ✅ |

### CODE REVIEW

File: python/sgl_jax/srt/models/qwen2.py [hunk 1]
```diff
@@ -281,9 +281,12 @@ def __call__(
         self,
         forward_batch: ForwardBatch,
         token_to_kv_pool: KVCache,
+        input_embeds: jax.Array | None = None,
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/models/qwen2.py]: can we put the input_embeds in forward_batch?

File: python/sgl_jax/srt/managers/scheduler.py [hunk 2]
```diff
@@ -668,6 +680,10 @@ def handle_generate_request(
         )
         req.tokenizer = self.tokenizer
 
+        if hasattr(recv_req, "mm_inputs") and recv_req.mm_inputs:
+            multimodal_embedding = recv_req.mm_inputs.get("multimodal_embedding")
+            req.mm_inputs = recv_req.mm_inputs
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/managers/scheduler.py]: can we just use mm_inputs?

File: python/sgl_jax/srt/multimodal/manager/mrope_utils.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
+
+
+def compute_mrope_positions(
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/multimodal/manager/mrope_utils.py:4]: why is this function not being used?

File: python/sgl_jax/srt/managers/scheduler.py (hunk 2 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/managers/scheduler.py]: we just use multimodal_embedding

File: python/sgl_jax/srt/models/qwen2.py (hunk 1 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/models/qwen2.py]: updated

File: python/sgl_jax/srt/models/qwen2.py [hunk 4]
```diff
@@ -281,9 +281,17 @@ def __call__(
...
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/models/qwen2.py]: please remove this parameter

**Review by pathfinder-pf — APPROVED** [review]

File: python/sgl_jax/srt/models/qwen2.py (hunk 4 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/models/qwen2.py]: ok
